hlib/python/hlib/frontend/relay_parser.py

```diff
@@ -0,0 +1,696 @@
+from tvm.relay.ty import TensorType, TupleType
+
+hcl.init(hcl.Float())
+
+debug_mode = False
+
+
+def get_model(model, shape):
+    """Gets the module (computation graph) and the parameters from the Keras model
+
+    Parameters
+    ----------
+    model : str
+        Path to model generated by Keras
+
+    shape : dict
+        Dictionary of input shapes into the model
+
+    Returns
+    -------
+    module : tvm.relay.Module
+        A relay module that contains the contents of the net
+
+    params : dict of str to NDArray
+        Parameters needed for the model
+    """
+    relay_model = keras.models.load_model(model)
+    module, params = relay_front.from_keras(relay_model, shape)
+    return module, params
+
+
+def gen_params(type_dict, env):
+    """Finds the parameters that we need to extract from the model
+
+    Parameters
+    ---------
+    type_dict: dict
+        the dictionary that contains the type of each variable in the environment
+
+    env: dict
+        the dictionary that contains the computational environment that sets up the
+        contained function {key: value:}
+
+    Returns
+    -------
+        a list of hcl.tensor.Tensor placeholders to hold the model params
+    """
+
+    params = []
+    for var in type_dict:
+        if (type_dict[var] == Var):
+            params.append(env[var])
+        elif (type_dict[var] == Tuple):
+            for item in env[var]:
+                if isinstance(item, hcl.tensor.Tensor):
+                    update_if(env, {item.name: item})
+    return params
+
+# tuple_extract
+
+
+def tuple_extract(tup, dict_t, env):
+    """takes a tuple and returns all the objects inside of it in a flat list
+
+    Parameters
+    ---------
+    tup: tuple
+        the tuple of objects we're trying to infer from
+
+    type_dict: dict
+        a dictionary of each object's type
+
+    env: dict
+        a dictionary of each object's computing environment
+
+    Returns
+    -------
+        a list of objects
+    """
+    result = []
+    if isPureList(tup):
+        for item in tup:
+            result.append(tuple_extract(item, dict_t, env))
+    else:
+        tp = dict_t[tup]
+        if tp == Var:
+            result.append(env[tup])
+        if tp == Tuple:
+            tup_env = env[tup]
+            result.append(tuple_extract(tup_env[1], tup_env[2], tup_env[3]))
+    return tuple(result)
+
+# let_bind (can be used for other frontends like Tensorflow, PyTorch)
+
+
+def let_bind(ntype, *arg):
+    """binds a computation to a variable
+
+    Parameters
+    ---------
+    ntype: tvm.relay-expr
+        the type of the binding
+
+    *arg: list of arguments
+        the arguments required for each ntype
+
+    Returns
+    -------
+        the object we bind to the output
+    """
+    if(debug_mode):
+        print("In bind")
+    if ntype == Call:
+        var = arg[0]
+        call_var = var[-1]
+        type_dict = arg[1]
+        call_type = type_dict[call_var]
```

Comment:
Move this line after ``params = arg[3]``. It makes more sense because we are extracting from ``arg`` first. Then, we use ``call_type`` later for the big ``if`` statement.